icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,404 @@ namespace icicle {
       MerkleProof& merkle_proof) const override
     {
-      ICICLE_LOG_DEBUG << "CPU CPUMerkleTreeBackend::get_merkle_proof() called for element index " << element_idx;
-      // Dummy implementation. TODO implement
-      uint64_t root = 123;
-      merkle_proof.allocate(
-        false /*pruned*/, element_idx, leaves + element_idx * get_leaf_element_size(), get_leaf_element_size(),
-        (const std::byte*)&root, sizeof(root));
-      return eIcicleError::SUCCESS; // TODO: Implement proof generation logic
+      ICICLE_ASSERT(m_tree_already_built) << "Tree cannot generate a merkle proof before built";
+      const uint64_t element_offset = leaf_idx * m_leaf_element_size;
+      const int l0_total_input_size = m_layers[0].m_hash.input_default_chunk_size();
+      ICICLE_ASSERT(leaf_idx < m_layers[0].m_nof_hashes * l0_total_input_size)
+        << "Element index out of range. Should be smaller than "
+        << m_layers[0].m_nof_hashes * l0_total_input_size / m_leaf_element_size;
+      uint64_t input_chunk_offset = (element_offset / l0_total_input_size) * l0_total_input_size;
+
+      // allocate merkle_proof memory
+      const auto [root, root_size] = get_merkle_root();
+      const auto input_chunk_size = m_layers[0].m_hash.input_default_chunk_size();
+      merkle_proof.allocate(is_pruned, leaf_idx, &leaves[input_chunk_offset], input_chunk_size, root, root_size);
+
+      std::byte* path = merkle_proof.allocate_path_and_get_ptr(is_pruned ? m_prooned_path_size : m_full_path_size);
+
+      // if not all leaves are stored
+      if (m_output_store_min_layer != 0) {
+        // Define a new tree tree to retrieve the forgotten hash results
+        const std::vector<Hash> sub_tree_layer_hashes(
+          m_layer_hashes.begin(), m_layer_hashes.begin() + m_output_store_min_layer + 1);
+        CPUMerkleTreeBackend sub_tree(sub_tree_layer_hashes, m_leaf_element_size, 0);
+
+        // build the sub tree
+        const uint64_t sub_tree_leaves_size =
+          l0_total_input_size * m_layers[0].m_nof_hashes / m_layers[m_output_store_min_layer].m_nof_hashes;
+        const uint64_t sub_tree_leaves_offset = (element_offset / sub_tree_leaves_size) * sub_tree_leaves_size;
+        const std::byte* sub_tree_leaves = &(leaves[sub_tree_leaves_offset]);
+        sub_tree.build(sub_tree_leaves, sub_tree_leaves_size, config);
+
+        // retrieve from the sub tree the path and increment path
+        const uint64_t sub_tree_input_chunk_offset = element_offset % sub_tree_leaves_size;
+        sub_tree.copy_to_path_from_store_min_layer(sub_tree_input_chunk_offset, is_pruned, path);
+      }
+
+      copy_to_path_from_store_min_layer(input_chunk_offset, is_pruned, path);
+      // print_proof(merkle_proof);
+      return eIcicleError::SUCCESS;
+    }
+
+    // Debug functions
+    eIcicleError print_tree(const std::byte* leaves, uint64_t leaves_size) const
+    {
+      ICICLE_ASSERT(m_tree_already_built) << "Tree cannot print before built";
```

Comment:
You can return an error instead of crashing in such cases

Reply:
thats a denbug function to be used by me - no one can access it